### PR TITLE
docs migration: Container Lifecycle Hooks into Concepts

### PR DESCRIPTION
**Pull request by chenopis**
Refactor [`docs/user-guide/container-environment.md`](https://kubernetes.io/docs/user-guide/container-environment) to `docs/concepts/containers/container-lifecycle-hooks.md` and `docs/concepts/containers/container-environment-variables.md`

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/2564)

### CODE REVIEW

**Comment by chenopis** [comment]
@steveperry-53 Ok, this is ready for you to review at your convenience. Thx

**Comment by steveperry-53** [comment]
@chenopis, This topic covers more that what the title suggests. We should either make the title more general or limit the topic content to Container lifecycle hooks. I vote for limiting the topic content to Container lifecycle hooks. If you agree, then I think you should remove the "Container environment" section, and shorten the topic overview.

**Comment by steveperry-53** [comment]
@chenopis, Capitalize API objects like Container. I don't think "hook" is an API object, so it shouldn't be capitalized.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+Services have dedicated IP addresses and are available to the container via DNS
+if [DNS addon](http://releases.k8s.io/{{page.githubbranch}}/cluster/addons/dns/) is enabled. 
+Because DNS is still not an enumerable protocol, 
+environment variables are provided so that Containers can do discovery.
+
+## Container hooks
+
+Container hooks provide information to the container about events in its management lifecycle.  
+For example, immediately after a container is started, it receives a `PostStart` hook.  
+Hooks are broadcast in the container with information about the lifecycle of the container.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: I think "with information about" is misleading. The hook handlers do not receive any arguments. When the Container's PostStart handler is called, it knows the Container was recently started; that's the only information it gets.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+if [DNS addon](http://releases.k8s.io/{{page.githubbranch}}/cluster/addons/dns/) is enabled. 
+Because DNS is still not an enumerable protocol, 
+environment variables are provided so that Containers can do discovery.
+
+## Container hooks
+
+Container hooks provide information to the container about events in its management lifecycle.  
+For example, immediately after a container is started, it receives a `PostStart` hook.  
+Hooks are broadcast in the container with information about the lifecycle of the container.  
+They are different from the events provided by Docker or other systems which are output by the container.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: I think these sentences about Docker and output events take the reader on an unnecessary side trip. I would remove them.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+This hook is sent immediately after a container is created and notifies the container that it has been created.  
+No parameters are passed to the handler. 
+It is not guaranteed that the hook will execute before the container entrypoint.
+
+`PreStop`
+
+This hook is called immediately before a container is terminated. 
+No parameters are passed to the handler. 
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: I think this would be more clear if we phrased it something like: "must complete before Kubernetes deletes the Container."

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+This hook is sent immediately after a container is created and notifies the container that it has been created.  
+No parameters are passed to the handler. 
+It is not guaranteed that the hook will execute before the container entrypoint.
+
+`PreStop`
+
+This hook is called immediately before a container is terminated. 
+No parameters are passed to the handler. 
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon. 
+The `SIGTERM` notification sent by Docker is also sent.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: I think we can remove this sentence about Docker sending SIGTERM. It strikes me as an implementation detail.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+No parameters are passed to the handler. 
+It is not guaranteed that the hook will execute before the container entrypoint.
+
+`PreStop`
+
+This hook is called immediately before a container is terminated. 
+No parameters are passed to the handler. 
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon. 
+The `SIGTERM` notification sent by Docker is also sent. 
+A more complete description of the termination behavior can be found in [Termination of Pods](/docs/user-guide/pods/#termination-of-pods).
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: Switch to active voice whenever you can. "You can find a more complete description ..."

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+### Hook handler execution
+
+When a management hook occurs,
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: This could be simpler. "Immediately after a Container starts, Kubernetes calls the Container's PostStart handler."

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+This event handler is blocking and must complete before the call to delete the container is sent to the Docker daemon. 
+The `SIGTERM` notification sent by Docker is also sent. 
+A more complete description of the termination behavior can be found in [Termination of Pods](/docs/user-guide/pods/#termination-of-pods).
+
+### Hook handler execution
+
+When a management hook occurs, 
+the management system calls into any registered hook handlers in the container for that hook.  
+These hook handler calls are synchronous in the context of the pod containing the container. 
+This means that for a `PostStart` hook,
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: This is hard to follow. "calls are synchronous ... This means ... asynchronously."

Consider something similar to the Discussion section in [Attaching Handlers to Container Lifecycle Events](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/).

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+If, for example, an http hook receiver is down and is unable to take traffic, 
+there is no attempt to resend. 
+In some rare cases, however, double delivery may occur.
+For instance, if a Kubelet restarts in the middle of sending a hook, 
+the hook may be resent after the Kubelet comes back up.
+
+
+### Hook handler implementations
+
+Hook handlers are the method by which Hooks are exposed to Containers.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: This first sentence is a bit contorted. I would remove it.

**Comment by steveperry-53** [comment]
@chenopis, Does this PR need all 7 files? I don't understand why volumes.md and images.md are here.

**Comment by steveperry-53** [comment]
@chenopis, Back to you.

**Comment by chenopis** [comment]
@steveperry-53 Ok, I'll break the topic into smaller parts and incorporate your feedback.Thanks!

**Comment by steveperry-53** [comment]
@chenopis, I don't think we need a separate topic for Container Environment. I think we could just remove that content and be fine. The part about FOO_SERVICE_HOST/PORT is obsolete (@pwittrock, correct?) and the rest is general information that I think is made clear throughout the docs.

**Comment by steveperry-53** [comment]
@chenopis, In What's next, add a link to [Attaching Handlers to Container Lifecycle Events](https://kubernetes.io/docs/tasks/configure-pod-container/attach-handler-lifecycle-event/).

**Comment by chenopis** [comment]
@steveperry-53 Ok, thanks. Will do!

**Comment by steveperry-53** [comment]
cc: @devin-donnelly for the moving of user-guide/images.md. My guess it that this file won't just get moved as a single topic under Concepts. It looks like some of the content should go under Setup Guides.

**Comment by chenopis** [comment]
@steveperry-53 Ok, I refactored it into two pages: container-lifecycle-hooks.md and container-environment.md. I also tracked down all the related links and updated them.

File: docs/concepts/workloads/containers/container-environment.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+---
+
+{% capture overview %}
+This page describes the resources available to containers in the Container Environment. 
+{% endcapture %}
+
+{:toc}
+
+{% capture body %}
+## Container environment
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-environment.md]: Use title casing: Container Environment.

File: docs/concepts/workloads/containers/container-environment.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+The Kubernetes Container Environment provides several important resources to containers:
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-environment.md]: I don't think Container Environment is the proper name of anything in Kubernetes. So no caps.

File: docs/concepts/workloads/containers/container-environment.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+* A filesystem, which is a combination of an [image](/docs/concepts/workloads/containers/images) and one or more [volumes](/docs/concepts/storage/volumes).
+* Information about the container itself.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-environment.md]: Throughout the doc, capitalize API objects like Container and Pod.

File: docs/concepts/workloads/containers/container-environment.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+
+```shell
+FOO_SERVICE_HOST=<the host the service is running on>
+FOO_SERVICE_PORT=<the port the service is running on>
+```
+
+Services have dedicated IP addresses and are available to the container via DNS,
+if [DNS addon](http://releases.k8s.io/{{page.githubbranch}}/cluster/addons/dns/) is enabled. 
+Because DNS is still not an enumerable protocol, however,
+environment variables are provided to enable Containers to do discovery.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-environment.md]: Try to avoid phrases like "still non an enumerable protocol." If you have to talk about the present moment as opposed to the future, isolate those statements a dedicated section at the end of the doc. Also, make sure any of these present/future statements are verified by a tech reviewer.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+and run code implemented in a handler when the corresponding lifecycle hook is executed.
+
+
+## Container hooks
+
+There are two hooks that are exposed to Containers:
+
+`PostStart`
+
+This hook executes immediately after a container is created.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: Caps for API objects like Containers and Pods.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+so it must complete before the call to delete the container can be sent. 
+No parameters are passed to the handler. 
+
+A more detailed description of the termination behavior can be found in
+[Termination of Pods](/docs/user-guide/pods/#termination-of-pods).
+
+### Hook handler implementations
+
+Containers can access a hook by implementing and registering a handler for that hook.
+There are two types of Hook handlers that can be implemented for Containers:
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: No caps for Hook.

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+
+Users should make their hook handlers as lightweight as possible.
+There are cases, however, when long running commands make sense,
+such as when saving state prior to stopping a container.
+
+### Hook delivery guarantees
+
+Hook delivery is intended to be *at least once*,
+which means that a hook may be called multiple times for any given event,
+such as for `start` or `stop`.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: start -> PostStart, stop -> PreStop

File: docs/concepts/workloads/containers/container-lifecycle-hooks.md [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+It is up to the hook implementation to handle this correctly.
+
+Generally, only single deliveries are made.
+If, for example, an HTTP hook receiver is down and is unable to take traffic,
+there is no attempt to resend.
+In some rare cases, however, double delivery may occur.
+For instance, if a kubelet restarts in the middle of sending a hook, 
+the hook may be resent after the kubelet comes back up.
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/container-lifecycle-hooks.md]: may -> might

File: docs/concepts/workloads/containers/images.md [hunk 17]
```diff
@@ -0,0 +1,328 @@
+---
+assignees:
+- erictune
+- thockin
+title: Images
+---
+
+{% capture overview %}
+
+Each container in a pod has its own image.  Currently, the only type of image supported is a [Docker Image](https://docs.docker.com/engine/tutorials/dockerimages/).
```
Comment (steveperry-53) [inline docs/concepts/workloads/containers/images.md]: I don't think Docker is the only supported image.

**Comment by steveperry-53** [comment]
@chenopis, Back to you. Comments in line.

**Comment by chenopis** [comment]
@steveperry-53 Ok, thanks. I've addressed all of your comments.

**Comment by chenopis** [comment]
@pwittrock Can you recommend an eng to do a tech review? Thx

**Comment by pwittrock** [comment]
@Random-Liu Would be a good person.

**Comment by chenopis** [comment]
Thanks @pwittrock

**Comment by chenopis** [comment]
@Random-Liu Do you have time to do a tech review for this PR?

**Comment by chenopis** [comment]
@Random-Liu Do you have time to review this? Thx
